### PR TITLE
Add CompOSE EOS table reader

### PR DESCRIPTION
**Pull request by nilsdeppe**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/IO/ComposeTable.cpp [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  parse_eos_quantities();
+  parse_eos_parameters();
+  parse_eos_table();
+}
+
+void ComposeTable::parse_eos_quantities() {
+  std::string line_buffer{};
+  const std::string filename{directory_to_read_from_ + "/eos.quantities"};
+  if (not file_system::check_if_file_exists(filename)) {
+    ERROR("File '" << filename << "' does not exist.");
```
Comment (kidder) [inline src/IO/ComposeTable.cpp:34]: test your ERRORs or exclude them from coverage

**Review by kidder — CHANGES_REQUESTED** [review]
squash

**Comment by nilsdeppe** [comment]
Added tests for the errors. I needed `file_system::copy` for this and then changed most of our file system library to use `std::filesystem`. This means there's less we need to maintain and should make us more resilient to OS, compiler, etc. switches

File: src/Utilities/FileSystem.cpp [hunk 2]
```diff
@@ -67,43 +69,21 @@ void create_directory(const std::string& dir, const double wait_time,
...
       Parallel::printf(
-          "create_directory: mkdir(%s) failed %d time(s). Error: %d\n", dir,
-          number_of_failures + 1, the_errno);
+          "create_directory: mkdir(%s) failed %d time(s). Error: %s\n", dir,
+          number_of_failures + 1, e.what());
       std::this_thread::sleep_for(std::chrono::duration<double>(
           wait_time *
           std::pow(wait_time_increase_factor, number_of_failures + 1)));
-      // LCOV_EXCL_STOP
+      // LCOV_EXCL_STOR
```
Comment (kidder) [inline src/Utilities/FileSystem.cpp]: ?

**Review by kidder — CHANGES_REQUESTED** [review]
the failing test does do file manipulation so you need to investigate...

**Review by kidder — APPROVED** [review]

**Pull request merged**